independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java

```diff
@@ -605,29 +608,53 @@ static ScopeInfo getValidScope(Collection<ScopeInfo> stereotypeScopes, Annotatio
+                }
+            }
+
+            // inherited stuff
+            ClassInfo aClass = beanClass;
+            Set<ClassInfo> scannedClasses = new HashSet<>();
+            while (aClass != null) {
+                if (!scannedClasses.add(aClass)) {
+                    continue;
+                }
+                Set<MethodInfo> methods = new HashSet<>();
+                for (MethodInfo method : aClass.methods()) {
+                    if (isOverriden(method, methods, index)) {
+                        continue;
                     }
-                } else if (annotationStore.hasAnnotation(method, DotNames.OBSERVES_ASYNC)) {
-                    // TODO observers are inherited
-                    asyncObserverMethods.add(method);
-                    if (!Modifier.isAbstract(beanClass.flags())) {
-                        // add only concrete classes
-                        beanClasses.add(beanClass);
-                        if (!hasBeanDefiningAnnotation) {
-                            LOGGER.debugf("Observer method found but %s has no bean defining annotation - using @Dependent",
-                                    beanClass);
+                    methods.add(method);
+                    if (annotationStore.getAnnotations(method).isEmpty()) {
+                        continue;
+                    }
+                    if (annotationStore.hasAnnotation(method, DotNames.OBSERVES)) {
+                        syncObserverMethods.computeIfAbsent(method, ignored -> new HashSet<>())
+                                .add(beanClass);
+                        if (!Modifier.isAbstract(beanClass.flags())) {
+                            // add only concrete classes
+                            beanClasses.add(beanClass);
+                            if (!hasBeanDefiningAnnotation) {
+                                LOGGER.debugf("Observer method found but %s has no bean defining annotation - using @Dependent",
+                                        beanClass);
+                            }
+                        }
+                    } else if (annotationStore.hasAnnotation(method, DotNames.OBSERVES_ASYNC)) {
+                        asyncObserverMethods.computeIfAbsent(method, ignored -> new HashSet<>())
+                                .add(beanClass);
+                        if (!Modifier.isAbstract(beanClass.flags())) {
+                            // add only concrete classes
+                            beanClasses.add(beanClass);
+                            if (!hasBeanDefiningAnnotation) {
+                                LOGGER.debugf("Observer method found but %s has no bean defining annotation - using @Dependent",
+                                        beanClass);
+                            }
                         }
                     }
                 }
+                Type superType = aClass.superClassType();
+                aClass = superType != null && superType.kind() == Type.Kind.CLASS
```

Comment:
1. End iteration if `superType.name()` is `java.lang.Object`
2. I don't think that `superType.kind() == Type.Kind.CLASS` is valid -> if `Foo extends Bar<String>` then you would not scan `Bar` as it has `Type.Kind.PARAMETERIZED_TYPE`
3. `index.getClassByName(superType.name())` may return null

Reply:
if `index.getClassByName(...)` returns `null`, the loop will not start the next iteration